Fix `extract_boundary_mesh()` for simplex meshes

QY-Shi opened the pull request:
Fix `GridGenerator::extract_boundary_mesh()` for simplex meshes by replacing `GeometryInfo<dim>::face_indices()` with cell->reference_cell().face_indices().

QY-Shi commented:
I have confirmed that this update fixes the issue for 2D simplex meshes, but the problem persists in 3D with a similar error.

I noticed that extract_boundary_mesh() explicitly iterates over 4 edges in 3D:
```c++
if (dim == 3)  
  for (unsigned int e = 0; e < 4; ++e)  
    {
      ...
    }
```

However, I am not fully familiar with deal.II's triangulation data structure and am unsure how this should be modified for simplex elements. Any guidance would be appreciated!

bangerth commented:
This looks correct to me. Would you be willing to add a test, perhaps a copy+modification of one of the `tests/codim_one/extract_boundary_mesh_*` tests?

You could also look at `tests/simplex/step-38.cc` which currently works around the issue in other ways (check the part that is guarded by `#ifdef USE_SIMPLEX`) and that could be simplified now with your patch.

bangerth commented:
For the 3d case, I'm willing to merge your patch here first if we had a test that demonstrates its functionality. We can then look at the other case separately.

drwells commented:
There are still a lot of things which need to be fixed in this function - e.g., the `swap_matrix` which determines the correct permutation of vertices to guarantee a consistently oriented surface.

luca-heltai commented:
> I have confirmed that this update fixes the issue for 2D simplex meshes, but the problem persists in 3D with a similar error.
> 
> I noticed that extract_boundary_mesh() explicitly iterates over 4 edges in 3D:
> 
> ```c++
> if (dim == 3)  
>   for (unsigned int e = 0; e < 4; ++e)  
>     {
>       ...
>     }
> ```
> 
> However, I am not fully familiar with deal.II's triangulation data structure and am unsure how this should be modified for simplex elements. Any guidance would be appreciated!

This should become

```
for(const auto e: cell->reference_cell()->line_indices()) {

...
}
```

bangerth reviewed (approved):
As mentioned in the comments, this is not yet correct. I don't know whether it's actually going to work, but I am certain that it's not going to make the problem any worse, so let's merge.

(I know for a fact that we are going to get tripped up on tet meshes because just a few lines below the changes here, we have `for (const unsigned int j : GeometryInfo<boundary_dim>::vertex_indices()) {... v_index = face->vertex_index(j);` which is going to fail for triangular faces.)

bangerth commented:
/rebuild

bangerth commented:
@QY-Shi Of course, we would be very happy if you had further improvements that fix the issue also on 3d meshes!

drwells commented on doc/news/changes/minor/20250201Shi:
```diff
@@ -0,0 +1,5 @@
+Fix: Replace GeometryInfo<dim>::face_indices() with
+cell->reference_cell().face_indices() in GridGenerator::extract_boundary_mesh()
+to correctly handle simplex meshes and prevent invalid face index access.
```
@bangerth I don't disagree with merging this but this changelog isn't right: there's still more work to do here. We don't yet correctly handle simplex meshes.

bangerth replied:
True. @QY-Shi Would you be willing to remove the changelog entry, or reword it?

QY-Shi commented:
> Would you be willing to remove the changelog entry, or reword it?

Sure! I removed the changeling file and squashed commit into one.

Apologies for not following up on this sooner. I haven't used this functionality since opening the PR, so I haven't looked deeper into the 3D cases. But I'm sure there are many more places like this that should be fixed, I'll find time to do that in the future.
